src/stream/src/executor/source/mod.rs

```diff
@@ -133,28 +133,42 @@ pub async fn apply_rate_limit(stream: BoxChunkSourceStream, rate_limit_rps: Opti
         )
     });
 
+    fn effective_chunk_size_of(chunk: &StreamChunk, limit: usize) -> usize {
+        let chunk_size = chunk.capacity();
+        if chunk_size == limit + 1 && chunk.ends_with_update() {
+            // If the chunk size exceed limit because of the last `UpdateDelete` operation,
+            // we should not split the chunk (actually we cannot with `split(limit)`).
+            chunk_size - 1
+        } else {
+            chunk_size
+        }
+    }
+
```

Comment:
This not accurate because we don't know how many `Update`s exist in the middle of the chunk.